### ISSUE TITLE
Taking in Condensed Distance Array for HBSCAN.fit()

### ISSUE THREAD

**Issue by windweller** [scikit-learn-contrib/hdbscan#125]
I'm trying to cluster around 80000 points, and a pair-wise distance matrix would be huge and painful to compute. `sklearn` and `fastcluster`'s `linkage` method takes in a condensed distance array and it saves me time to only compute half of the distance matrix (since it's symmetric).

```python
distance_matrix = pairwise_distances(blobs)
clusterer = hdbscan.HDBSCAN(metric='precomputed')
clusterer.fit(distance_matrix)
```

Does HBSCAN take in a condensed array that's the upper triangular area of the pairwise distance matrix?

**Comment by lmcinnes** [comment]
Not at the current time; if you can specify the data as vectors and provide
a metric that is supported by BallTrees (a reasonably wide range) then
HDBSCAN will never actually compute the full distance matrix and require
significantly less memory. another alternative is to make a sparse distance
matrix (truncate out the large distances) and use that instead.

**Comment by windweller** [comment]
Thank you Imcinnes! I'm handling text clustering. Pair-wise distance like edit distance (Levenshtein distance) is most helpful, but hard to vectorize. Vectorization can result in high-dimensional vectors (which I believe is hard for DBSCAN-type clustering to handle?)

Can you say a bit more about the sparse distance matrix? Condensed distance matrix is just flattening out the upper triangular part of distance matrix, what is sparse distance matrix? Thank you!

**Comment by lmcinnes** [comment]
Text with Levenstein is definitely an issue. To manage the sparse matrix
version you want to compute only distances between relatively similar
pairs, and then build a sparse matrix (see scipy.sparse documentation) that
has values for those pairs; any sparse entries (other than the main
diagonal) will be assumed to have effectively infinite distance. This is
good enough as long as you compute enough entries of the matrix.

**Comment by windweller** [comment]
It seems like sparse matrix is the way to go! Endless thanks!
Do you have any recommendation on whether to use row/column sparse matrix? (It's symmetric for text pairs anyway).

**Comment by lmcinnes** [comment]
For building the matrix itself I would use COO format, and then convert
that to CSR when you are ready to pass it to clustering. Fortunately
scipy.sparse makes the conversion trivial.

**Comment by windweller** [comment]
Thank you :) this has been very helpful!

**Comment by lmcinnes** [comment]
Glad that I was able to help.
